File: Algorithm/Algorithm/1062.swift
////
////  main.swift
////  Algorithm
////
////  Created by 정성훈 on 2021/03/11.
////
//
import Foundation

extension String {
    subscript(offset: Int) -> Character {
        self[index(startIndex, offsetBy: offset)]
    }
}

var funcCount = 0
func combination(originalPool: [Character], charPool: inout Set<Character>, startIndex: Int, answer: inout Int) {
    funcCount += 1
    if charPool.count >= K {
        let wordCount = canReadWordCount(charPool: &charPool)
        if wordCount > answer {
            answer = wordCount
        }
        return
    } else {
        for i in startIndex..<originalPool.count {
            charPool.insert(originalPool[i])
            combination(originalPool: originalPool, charPool: &charPool, startIndex: i + 1, answer: &answer)
            charPool.remove(originalPool[i])
        }
        return
    }
}

var loopCount = 0
func canReadWordCount(charPool: inout Set<Character>) -> Int {
    var count = 0
    wordLoop: for word in words { // global variable
        for char in word[word.index(word.startIndex, offsetBy: 4)...word.index(word.endIndex, offsetBy: -5)] {
            loopCount += 1
            if charPool.contains(char) == false {
              continue wordLoop
            }
        }
        count += 1
    }
    return count
}

func makeOriginalPool(words: [String]) -> Array<Character> {
    var originalPool: Array<Character> = ["a", "c", "i", "n", "t"]

    var pool = Set(originalPool)
    for word in words {
        for char in word {
            if pool.contains(char) == false {
                pool.insert(char)
                originalPool.append(char)
            }
        }
    }

    return originalPool
}

var input = readLine()!.components(separatedBy: " ").map{ Int($0)! }
var N = input[0]
var K = input[1]
var words = [String]()

for _ in 0..<N {
    words.append(readLine()!)
}

var answer = 0

if K >= 5 {
    let originalPool = makeOriginalPool(words: words)
    if originalPool.count > K {
        var charPool: Set<Character> = ["a", "c", "i", "n", "t"]
        combination(originalPool: originalPool, charPool: &charPool, startIndex: 5, answer: &answer)
    }
}

print(answer)

File: Algorithm/Algorithm/1062.cpp
#include <iostream>
#include <set>
#include <vector>
#include <string>
using namespace std;

int N, K;
vector<string> words;

void combination(const vector<char>& originalPool, set<char>& charPool, int startIndex, int& answer);
int canReadCount(set<char>& charPool);
vector<char> makeOriginalPool(const vector<string>& words);

int main(void)
{
    cin >> N;
    cin >> K;
    for(int i = 0; i < N; i++)
    {
        string input;
        cin >> input;
        words.push_back(input);
    }
    
    int answer = 0;
    if(K >= 5)
    {
        vector<char> originalPool = makeOriginalPool(words);
        if(originalPool.size() > K)
        {
            set<char> charPool = {'a', 'c', 'i', 'n', 't'};
            combination(originalPool, charPool, 5, answer);
        } else {
            answer = words.size();
        }
    }
    
    cout << answer;
    return 0;
}

void combination(const vector<char>& originalPool, set<char>& charPool, int startIndex, int& answer)
{
    if (charPool.size() >= K)
    {
        const int wordCount = canReadCount(charPool);
        if (wordCount > answer)
        {
            answer = wordCount;
        }
        return;
    }
    else
    {
        for(int i = startIndex, size = originalPool.size(); i < size; i++)
        {
            charPool.insert(originalPool[i]);
            combination(originalPool, charPool, i + 1, answer);
            charPool.erase(originalPool[i]);
        }
        return;
    }
    
}

int canReadCount(set<char>& charPool)
{
    int count = 0;
    for(auto iter = words.begin(), end = words.end(); iter != end; iter++)
    {
        const string& word = (*iter);
        bool canRead = true;
        for(int i = 4, len = word.length() - 4; i < len; i++)
        {
            if(charPool.find(word[i]) == charPool.end())
            {
                canRead = false;
                break;
            }
        }
        if(canRead)
        {
            count += 1;
        }
    }
    return count;
}

vector<char> makeOriginalPool(const vector<string>& words)
{
    vector<char> originalPool = {'a', 'c', 'i', 'n', 't'};
    
    set<char> pool = {'a', 'c', 'i', 'n', 't'};
    for(auto iter = words.begin(), end = words.end(); iter != end; iter++)
    {
        const string& word = (*iter);
        for(int i = 0, len = word.length(); i < len; i++)
        {
            if(pool.find(word[i]) == pool.end())
            {
                pool.insert(word[i]);
                originalPool.push_back(word[i]);
            }
        }
    }
    return originalPool;
}
